batch.js

```diff
@@ -0,0 +1,317 @@
+              // isPgCreatePayloadType: true,
+              pgIntrospection: table,
+            },
+          )
+
+          // Create upsert fields from each introspected table
+          const fieldName = `upsert${tableTypeName}Batch`
+          memo = build.extend(
+            memo,
+            {
+              [fieldName]: fieldWithHooks(
+                fieldName,
+                context => {
+                  const { getDataFromParsedResolveInfoFragment } = context
+                  return {
+                    description: `Upserts a batch of \`${inflection.pluralize(
+                      tableTypeName,
+                    )}\`.`,
+                    type: PayloadType,
+                    args: {
+                      input: {
+                        type: new GraphQLNonNull(InputType),
+                      },
+                    },
+                    async resolve(data, { input }, { pgClient }, resolveInfo) {
+                      const parsedResolveInfoFragment = parseResolveInfo(
+                        resolveInfo,
+                      )
+                      const resolveData = getDataFromParsedResolveInfoFragment(
+                        parsedResolveInfoFragment,
+                        PayloadType,
+                      )
+                      const upsertedRowsAlias = sql.identifier(Symbol())
+                      const query = queryFromResolveData(
+                        upsertedRowsAlias,
+                        upsertedRowsAlias,
+                        resolveData,
+                        {},
+                      )
+                      const inputData = input[inputFieldName]
+
+                      // no inputs, no upserts
+                      if (!inputData || !inputData.length) {
+                        return {
+                          clientMutationId: input.clientMutationId,
+                          data: [],
+                        }
+                      }
+
+                      // Store attributes (columns) for easy access
+                      const attributes = pgIntrospectionResultsByKind.attribute
+                        .filter(attr => attr.classId === table.id)
+                        .filter(attr => pgColumnFilter(attr, build, context))
+                        .filter(attr => !omit(attr, 'create'))
+
+                      // Figure out the pkey constraint
+                      const primaryKeyConstraint = pgIntrospectionResultsByKind.constraint
+                        .filter(con => con.classId === table.id)
+                        .filter(con => con.type === 'p')[0]
+
+                      // Figure out to which column that pkey constraint belongs to
+                      const primaryKeys =
+                        primaryKeyConstraint &&
+                        primaryKeyConstraint.keyAttributeNums.map(
+                          num => attributes.filter(attr => attr.num === num)[0],
+                        )
+
```

Comment:
The three statements above should be precomputed outside the resolver for performance reasons.